### ISSUE TITLE
DeepCreamPyV2 License Discussion

### ISSUE THREAD

**Issue by deeppomf** [deeppomf/DeepCreamPy#154]
With the release of DCPv2 soon, we need to discuss its license and why I want to move away from [AGPL](https://www.gnu.org/licenses/agpl-3.0.en.html).

What has changed since I decided to use AGPL back in October 2018 (Issue #42)?

Because of financial difficulties, I need to support myself in part off DCPv2.
That alone wouldn't make me switch away from AGPL.

What does motivate switching are online clones of DCP where users can upload images to a website and have them be decensored by a copy of DCP running on the website's server. There already is an online clone of DCPv1 so this isn't speculation. AGPL does not prevent online clones; it only requires they share their source code under the same license.

_Since these web-based clones use DCP to decensor, my releases of DCPv2 would be forced to compete against itself should I continue using AGPL._

Furthermore, web-based clones don't contribute back to the development of DCP. The online clone of DCPv1 shared their source code as per AGPL requirements, but their source code is almost identical to DCPv1 and has nothing useful to add. The author of the online clone offered to share the data of uploaded images with me, but that information is not beneficial for DCP either.

What license will DCPv2 switch to? Don't know.

tl;dr: Because AGPL doesn't prevent web-based clones of DCP, I want to switch away from AGPL.

Please leave your thoughts in the comments.

**Comment by Lucretious** [comment]
Hi, I do not know the differences between these licenses, but I think, you should just choose the license  that you want, or need. It is your project, so you have the choice. 
But if I remember correctly, somebody said in the other thread, that a license change could mean problems, because you use code already under license, and you are not allowed to use that anymore, if you switch...? Sorry, I am not a programmer, but somebody who will simply buy and use your finished product. 

So, from a enduser point of view: it really doesn´t matter to me, as long, as you do not illegally use code from somebody else. Other then that, please choose, what suits your needs. :-)

Oh, and do you maybe have a rough estimation when DCP 2 might be ready?

**Comment by noirscape** [comment]
I would like to note/state that for a license change you need consent from
all existing contributors who have substantive contributions to the code or
you need to pull all external contributions from your code.

Also, as far as relicensing goes, DCPv1 will always be under the AGPL, even
if you relicense it later (this is called estoppel.)

Fwiw, IANAL.

**Comment by ccppoo** [comment]
for financial problems, I suggest releasing as a software product.
https://github.com/deeppomf/DeepCreamPy/issues/117 also mentioned before.
Before choosing licenses, how about choosing whether DCPv2 closed or open l? you could change to open source after some support like Swift or CentOS did

**Comment by deeppomf** [comment]
@noirscape I'll make sure DCPv2 is all my code.
@ccppoo Still undecided on the license or the financial model.

**Comment by ccppoo** [comment]
agony between absence of progressive or brilliant PRs, in case of opening your code in github (whatever license you choose), 
and finding financial model at the sametime 
(it looks like that for me) seems to have problem.

I would rather say like decide...
1. non-paid <-> donation(patreon or sth)<-> sell as product
then
2. open source <-> closed source
then
3.  choosing license what satisfy above

hope you could find one that just fit to your needs

**Comment by DarkPulse** [comment]
There's definitely good reasons for switching the code to a different source model.

For one, as you've said, you definitely don't want to compete against yourself, and if anyone can just take your code, slap some new models in, and offer it for free, that's a disaster for you, especially if you're heavily depending on the project to make ends' meet.

For two, you said in the past thread (about monetization) that you do intend to release the source code once you're through with the project entirely (or the 2.X branch) and that's also probably worth noting - this is less of a "won't ever have source released" prospect versus that when it's exceeded its useful life, you'll make it available to others. This is less a case of "DCPv2 will never have source" as opposed to "DCPv2 will be closed-source for the immediate future."

Three, there's some licenses that do allow you to add in your own clauses. For example, you could make the source code available upon request, but with conditions prohibiting commercial use, or requiring a watermark that would direct towards your version for going without watermarks, or stuff like I suggested in the monetization thread about essentially producing inferior copies after awhile or whatever.

And lastly, there's also the fact that in a way, you'll probably compete against someone else's efforts no matter what. DCPv1 is out there and available for download; when DCPv2 comes out I'm sure that will cause some people with coding knowledge to try to port and/or improve DCPv1 to be much more close to it.

I really can't say what license you should go for, but I'd definitely think the best thing for you to do will be to make it clear that it wll **NOT** stay closed-source forever, that buying this supports you more directly in your time of need, that DCPv1 will always be free if anyone does want to try to improve it (until DCPv2 is similarly sourced), so all you can do is hope that whatever you charge for it and explain your reasons for charging and keeping it temporarily closed-source are acceptable to people.

If not, you at least tried.

**Comment by deeppomf** [comment]
I decided DCPv2 will be supported through Patreon or Paypal. DCPv2 might be sold as a product for the Japanese translation.
DCPv2 source code will be available with restrictions (e.g. no commercial use). DCP has benefited immensely from contributions from others, and it would be a shame if this repo dies off.
DCPv2 blows DCPv1 out of the water. I'm confident if there will to be competition it would not be from a fork of DCPv1.

@DarkPulse 
Is there a way to talk to you privately about something?
I have a Twitter and Reddit account.

**Comment by DarkPulse** [comment]
@deeppomf I've got both. Which would you prefer?

As for funding, Patreon or Paypal would be good, and commercial use is a no-brainer (though realistically, anyone who's doing commercial releases of the work could probably pay an artist for a manual decensor).

**Comment by deeppomf** [comment]
Twitter

**Comment by deeppomf** [comment]
@DarkPulse Twitter

**Comment by DarkPulse** [comment]
@deeppomf Sent a message; Lurch says hello.

**Comment by miniBill** [comment]
Well, if you make people sign CLAs you can release the code as cc-by-nc-sa *and* keep the version you distribute commercially as closed source.

This would meet points 1, 2, 3 and 4.
You can make money by:
* asking for donations (which you already mentioned);
* selling additional features / non-public training data (having the code as OSS and the data as commercial/proprietary is for example what happened to old quake software);
* selling commercial licenses;
* selling translated versions;
* selling it "as a service" with some kind of per-upload fee;
* ...

**Comment by rekaXua** [comment]
Hmmm... I'm not a lawyer, but are you sure, that "deeppomf" that is in EULA can be a legal subject?

**Comment by n-bernat** [comment]
> Well, if you make people sign CLAs you can release the code as cc-by-nc-sa _and_ keep the version you distribute commercially as closed source.

Creative Commons licenses are not designed for software.

https://creativecommons.org/faq/#can-i-apply-a-creative-commons-license-to-software

**Comment by deeppomf** [comment]
@rekaXua It's that or I give my real name.
If there's a legal problem, I'll cross that bridge when I get there.
